### PR TITLE
Use FxCop Analyzers instead of Legacy Analyzers

### PR DESCRIPTION
**Pull request by WardenGnaw**
This PR disables the Legacy FxCop and uses FxCop Analyzers instead. 

Due to the upgrade, there are many issues that have been caught but are disabled in `IDECodeAnalysis.ruleset`. 
I have added the description of the rules we have disabled.

Let me know if there are ones I should fix instead of having them disabled for now.

Each commit is a fix for a rule except the first and last commit.

### CODE REVIEW

File: src/IDECodeAnalysis.ruleset [hunk 1]
```diff
@@ -1,52 +1,95 @@
...
+    <Rule Id="CA2141" Action="Error" Description="CA2141:Transparent methods must not satisfy LinkDemands" />
+    <Rule Id="CA2142" Action="Error" Description="CA2142: Transparent code should not be protected with LinkDemands" />
+    <Rule Id="CA2143" Action="Error" Description="CA2143: Transparent methods should not use security demands" />
+    <Rule Id="CA2144" Action="Error" Description="CA2144: Transparent code should not load assemblies from byte arrays" />
+    <Rule Id="CA2145" Action="Error" Description="CA2145: Transparent methods should not be decorated with the SuppressUnmanagedCodeSecurityAttribute" />
+    <Rule Id="CA2146" Action="Error" Description="CA2146: Types must be at least as critical as their base types and interfaces" />
+    <Rule Id="CA2147" Action="Error" Description="CA2147: Transparent methods may not use security asserts" />
+    <Rule Id="CA2149" Action="Error" Description="CA2149: Transparent methods must not call into native code" />
+
+    <!-- TODO(waan): Enable errors and warnings below -->
```
Comment (gregg-miskelly) [inline src/IDECodeAnalysis.ruleset:52]: >Enable [](start = 21, length = 6)

Note: Some of these (ex: CA1031, CA1707) are things we will never enable.

File: src/DebugEngineHost/HostLogger.cs [hunk 2]
```diff
@@ -83,7 +83,7 @@ internal static StreamWriter GetStreamForName(string logFileName, bool throwInUs
             }
             else
             {
-                throw new ArgumentException("logFileName");
+                throw new ArgumentException(null, nameof(logFileName));
```
Comment (gregg-miskelly) [inline src/DebugEngineHost/HostLogger.cs:86]: >ArgumentException [](start = 26, length = 17)

use `ArgumentOutOfRangeException`

File: src/MIDebugEngine/AD7.Impl/AD7Engine.cs [hunk 3]
```diff
@@ -199,7 +194,7 @@ public int Attach(IDebugProgram2[] portProgramArray, IDebugProgramNode2[] progra
             if (celtPrograms != 1)
             {
                 Debug.Fail("SampleEngine only expects to see one program in a process");
-                throw new ArgumentException();
+                throw new ArgumentException(null, nameof(celtPrograms));
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7Engine.cs:197]: >ArgumentException [](start = 26, length = 17)

Same

File: src/MIDebugEngine/AD7.Impl/AD7Engine.cs [hunk 4]
```diff
@@ -220,7 +215,7 @@ public int Attach(IDebugProgram2[] portProgramArray, IDebugProgramNode2[] progra
                     if (processId.ProcessIdType != (uint)enum_AD_PROCESS_ID.AD_PROCESS_ID_SYSTEM)
                     {
                         Debug.Fail("Invalid process to attach to");
-                        throw new ArgumentException();
+                        throw new ArgumentException(nameof(processId));
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7Engine.cs:218]: >ArgumentException [](start = 34, length = 17)

Same

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 5]
```diff
@@ -243,7 +243,7 @@ public DebuggedProcess(bool bLaunched, LaunchOptions launchOptions, ISampleEngin
             }
             else
             {
-                throw new ArgumentOutOfRangeException("LaunchInfo.options");
+                throw new ArgumentOutOfRangeException(null, nameof(_launchOptions));
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs:246]: >(null,  [](start = 53, length = 7)

Remove the `null`. You want to use the single-argument constructor.

File: src/OpenDebugAD7/Document.cs [hunk 6]
```diff
@@ -86,7 +86,7 @@ public byte[] GetChecksums(Guid guidHashAlgorithm)
             }
             else
             {
-                throw new ArgumentException("guidHashAlgorithm");
+                throw new ArgumentException(null, nameof(guidHashAlgorithm));
```
Comment (gregg-miskelly) [inline src/OpenDebugAD7/Document.cs:89]: > ArgumentException(null,  [](start = 25, length = 25)

`ArgumentOutOfRangeException`

**Review by gregg-miskelly — APPROVED** [review]
Otherwise LGTM

File: src/MICore/Transports/TcpTransport.cs [hunk 7]
```diff
@@ -61,7 +61,7 @@ public override void InitStreams(LaunchOptions options, out StreamReader reader,
                     null /*UserCertificateSelectionCallback */
                     );
 
-                sslStream.AuthenticateAsClientAsync(tcpOptions.Hostname, certStore.Certificates, System.Security.Authentication.SslProtocols.Tls, false /* checkCertificateRevocation */).Wait();
+                sslStream.AuthenticateAsClientAsync(tcpOptions.Hostname, certStore.Certificates, System.Security.Authentication.SslProtocols.None, false /* checkCertificateRevocation */).Wait();
```
Comment (WardenGnaw) [inline src/MICore/Transports/TcpTransport.cs:64]: Is changing from Tls to None ok? It was requiring it to make it determined by the OS.

**Review by gregg-miskelly — APPROVED** [review]

File: src/MICore/Transports/TcpTransport.cs (hunk 7 above)
Comment (pieandcakes) [reply src/MICore/Transports/TcpTransport.cs:64]: Starting with .NET Framework 4.7, this method authenticates using None, which allows the operating system to choose the best protocol to use, and to block protocols that are not secure. In .NET Framework 4.6 (and .NET Framework 4.5 with the latest security patches installed), the allowed TLS/SSL protocols versions are 1.2, 1.1, and 1.0 (unless you disable strong cryptography by editing the Windows Registry).

Comment (pieandcakes) [reply src/MICore/Transports/TcpTransport.cs:64]: I'd #def and say when you are using 4.7.2 (IsXplat false?) then use none.

**Pull request merged**